
- 오류가 발생하게 되면, 오류는 클라이언트에서 가장 가까운 `error.js` 또는 `<Suspense>` 측에서 catch된다.
	- 이 때, UI에서 처리되도록 오류를 반환하기 위해서 `try/catch`문을 사용하는 것이 권장된다고 한다.

- 예를 들어, Server Action에서 새로운 항목을 생성하는 중에 오류가 발생하면 아래와 같이 메시지를 반환하여 처리할 수 있다.

> `<ClientComponent/>`
```tsx
"use client";

import React from "react";
import { createTodo, createUser } from "../utils/action/actions";
import { useFormState } from "react-dom";

// 메시지 관리용 state
const initialState:
  | undefined
  | {
      status: number;
      message: string;
    } = undefined;

const ClientComponent = () => {
  // useFormState hook을 사용하여, server action함수와, 초기값을 전달
  const [state, formAction] = useFormState(createTodo, initialState);

  return (
    <form action={formAction}>
      <label htmlFor="entry">할 일</label>
      <input type="text" id="entry" name="entry" />
      {/* ... */}
      <p aria-live="polite" className="sr-only">
        {state ? `${state.message} (${state.status})` : ""}
      </p>
      <button>추가</button>
    </form>
  );
};

export default ClientComponent;
```

> `action.ts`
```ts
"use server";
import { schema as entrySchema } from "../schema/textareaSchema";

export async function createTodo(prevState: any, formData: FormData) {
  try {
    const validatedEntryFields = entrySchema.safeParse({
      entry: formData.get("entry"),
    });

    // Return early if the form data is invalid
    if (!validatedEntryFields.success) {
      throw new Error("등록 과정에서 입력된 데이터가 없음을 확인했습니다.");
    }

    // Mutate data
    return { status: 200, message: "등록이 완료되었습니다." };
  } catch (e: any) {
    if (e.message === "등록 과정에서 입력된 데이터가 없음을 확인했습니다.") {
      return {
        status: 400,
        message: e.message,
      };
    } else {
      // ...
      return {
        status: 400,
        message: e.message,
      };
    }
  }
}
```

>`textareaSchema.ts`
```ts
import { z } from "zod";

export const schema = z.object({
  entry: z.string().refine((value) => value.length !== 0, {
    message: "Invalid entry",
  }),
});

```